packages/cozy-harvest-lib/src/components/OAuthService.js

```diff
@@ -0,0 +1,330 @@
+        if (result?.type !== 'dismiss' && result?.type !== 'cancel') {
+          // @ts-ignore wrong return type associated to webviewIntent.call
+          resolve({ result: 'ERROR', error: result })
+          return
+        }
+
+        webviewIntent.call('closeInAppBrowser')
+        resolve({ result: 'CLOSED' })
+        return
+      } catch (err) {
+        webviewIntent.call('closeInAppBrowser')
+        resolve({ result: 'ERROR', error: err })
+        return
+      }
+    }
+    doExec()
+  })
+}
+
+/**
+ * @typedef IntentsApi
+ * @property {Function} fetchSessionCode
+ * @property {Function} showInAppBrowser
+ * @property {Function} closeInAppBrowser
+ * @property {String} tokenParamName
+ */
+
+/**
+ * Opens given url in the intentsApi inAppBrowser
+ *
+ * @param {Object} options
+ * @param {String} options.oAuthUrl - final url to open in the intentsApi inAppBrowser
+ * @param {IntentsApi} options.intentsApi - WebviewService object to use
+ * @param {Function} options.registerRealtime
+ * @returns {Promise<OAuthServiceResult>}
+ */
+function openIntentsApiInAppBrowser({
+  oAuthUrl,
+  registerRealtime,
+  intentsApi
+}) {
+  return new Promise((resolve, reject) => {
+    const doExec = async () => {
+      const {
+        fetchSessionCode,
+        showInAppBrowser,
+        closeInAppBrowser,
+        tokenParamName = 'session_code'
+      } = intentsApi
+
+      const isReady = Boolean(
+        intentsApi?.fetchSessionCode &&
+          intentsApi?.showInAppBrowser &&
+          intentsApi?.closeInAppBrowser
+      )
+
+      if (!isReady) {
+        return reject('Missing method in intentsApi')
+      }
+      try {
+        registerRealtime(data => {
+          closeInAppBrowser()
+          return resolve(data)
+        })
+        const sessionCode = await fetchSessionCode()
+        logger.debug('got session code', sessionCode)
+        const iabUrl = new URL(oAuthUrl)
+        iabUrl.searchParams.append(tokenParamName, sessionCode)
+        // we need to decodeURIComponent since toString() encodes URL
+        // but native browser will also encode them.
+        const urlToOpen = decodeURIComponent(iabUrl.toString())
+        logger.debug('url to open: ', urlToOpen)
+        const result = await showInAppBrowser(urlToOpen)
+        if (result?.state !== 'dismiss' && result?.state !== 'cancel') {
+          return resolve({ result: 'ERROR', error: result })
+        }
+        return resolve({ result: 'CLOSED' })
+      } catch (err) {
+        return resolve({ result: 'ERROR', error: err })
+      }
+    }
+    doExec()
+  })
+}
+
+/**
+ * Register to realtime events
+ *
+ * @param {Object} options
+ * @param {Object} options.client
+ * @param {import('cozy-client/types/types').KonnectorsDoctype} options.konnector
+ * @returns  {Function}
+ */
+function registerRealtime({ client, konnector }) {
+  return resolve => {
+    const realtime = new CozyRealtime({ client })
```

Comment:
Can we use the realtime plugin from the client here? In order to be sure to have a singleton and not many opened WS?